services/gitlab/gitlab-coverage.service.js

```diff
@@ -0,0 +1,121 @@
+const { optionalUrl } = require('../validators')
+const { BaseSvgScrapingService, NotFound, redirector } = require('..')
+
+const badgeSchema = Joi.object({
+  message: Joi.alternatives()
+    .try(
+      Joi.string().regex(/^([0-9]+\.[0-9]+%)|unknown$/),
+      Joi.equal('unknown')
+    )
+    .required(),
+}).required()
+
+const queryParamSchema = Joi.object({
+  gitlab_url: optionalUrl,
+}).required()
+
+const documentation = `
+<p>
+  Important: If your project is publicly visible, but the badge is like this:
+  <img src="https://img.shields.io/badge/build-not&nbsp;found-red" alt="build not found"/>
+</p>
+<p>
+  Check if your pipelines are publicly visible as well.<br />
+  Navigate to your project settings on GitLab and choose General Pipelines under CI/CD.<br />
+  Then tick the setting Public pipelines.
+</p>
+<p>
+  Now your settings should look like this:
+</p>
+<img src="https://user-images.githubusercontent.com/12065866/67156911-e225a180-f324-11e9-93ad-10aafbb3e69e.png" alt="Setting Public pipelines set"/>
+<p>
+Also make sure you have set up code covrage parsing as described <a href="https://docs.gitlab.com/ee/ci/pipelines/settings.html#test-coverage-parsing">here</a>
+</p>
+<p>
+  Your badge should be working fine now.
+</p>
+`
+
+class GitlabCoverage extends BaseSvgScrapingService {
+  static get category() {
+    return 'coverage'
+  }
+
+  static get route() {
+    return {
+      base: 'gitlab/coverage',
+      pattern: ':user/:repo/:branch+',
+      queryParamSchema,
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Gitlab code coverage',
+        namedParams: {
+          user: 'gitlab-org',
+          repo: 'gitlab-runner',
+          branch: 'master',
+        },
+        staticPreview: this.render({ percentage: 100 }),
+        documentation,
+      },
+      {
+        title: 'Gitlab code coverage (self-hosted)',
+        namedParams: { user: 'GNOME', repo: 'libhandy', branch: 'master' },
+        queryParams: { gitlab_url: 'https://gitlab.gnome.org' },
+        staticPreview: this.render({ percentage: 50 }),
+        documentation,
+      },
+    ]
+  }
+
+  static render({ percentage }) {
+    return {
+      message: `${percentage}%`,
+      color: coveragePercentageColor(percentage),
+    }
+  }
+
+  async handle(
+    { user, repo, branch = 'master' },
+    { gitlab_url: baseUrl = 'https://gitlab.com' }
+  ) {
+    const { message: percentageText } = await this._requestSvg({
+      schema: badgeSchema,
+      url: `${baseUrl}/${user}/${repo}/badges/${branch}/coverage.svg?style=flat-square`,
+      errorMessages: {
+        401: 'repo not found',
+        404: 'repo not found',
+      },
+    })
+    if (!/[0-9.]+/.test(percentageText)) {
+      throw new NotFound({ prettyMessage: 'not set up' })
+    }
+    // Parse percentage to a number
+    const percentage = Number(/[0-9.]+/.exec(percentageText)[0]).toFixed(0)
+    return this.constructor.render({ percentage })
```

Comment:
Think this could be expressed a bit more simply, and leave the precision decision to the render function where it's better addressed. Note the arg rename for the render function to match the suggested param name change above. 

The service class for Codecov is a good example to reference, as it also uses the upstream Svg https://github.com/badges/shields/blob/master/services/codecov/codecov.service.js 

```suggestion
    return this.constructor.render({ coverage: Number(percentageText.slice(0, -1)) })
```

Reply:
[Done](https://github.com/badges/shields/pull/5262/commits/2fd761890095c18e503140a0f9fe2524eacd2c9d#diff-481a722248c7617406848306c7335ce8R97)